akka-bench-jmh/src/main/scala/akka/actor/BenchmarkActors.scala

```diff
@@ -0,0 +1,163 @@
+    val latch = new CountDownLatch(numPairs * 2)
+    val actors = for {
+      i <- (1 to numPairs).toVector
+    } yield {
+      val ping = system.actorOf(PingPong.props(messagesPerPair, latch).withDispatcher(fullPathToDispatcher))
+      val pong = system.actorOf(PingPong.props(messagesPerPair, latch).withDispatcher(fullPathToDispatcher))
+      (ping, pong)
+    }
+    (actors, latch)
+  }
+
+  private def startUserServiceActors(numActors: Int, numQueriesPerActor: Int, numUsersInDBPerActor: Int, dispatcher: String, randomGen: Random)(implicit system: ActorSystem) = {
+    val fullPathToDispatcher = "akka.actor." + dispatcher
+    val latch = new CountDownLatch(numActors)
+    val actors = for {
+      i <- (1 to numActors).toVector
+    } yield system.actorOf(UserServiceActor.props(numQueriesPerActor, numUsersInDBPerActor, latch, randomGen).withDispatcher(fullPathToDispatcher))
+    (actors, latch)
+  }
+
+  private def initiatePingPongForPairs(refs: Vector[(ActorRef, ActorRef)], inFlight: Int) = {
+    for {
+      (ping, pong) <- refs
+      _ <- 1 to inFlight
+    } {
+      ping.tell(Message, pong)
+    }
+  }
+
+  private def printProgress(totalMessages: Long, numActors: Int, startNanoTime: Long) = {
+    val durationMicros = (System.nanoTime() - startNanoTime) / 1000
+    println(f"  $totalMessages messages by $numActors actors took ${durationMicros / 1000} ms, " +
+      f"${totalMessages.toDouble / durationMicros}%,.2f M msg/s")
+  }
+
+  def requireRightNumberOfCores(numCores: Int) =
+    require(
+      Runtime.getRuntime.availableProcessors == numCores,
+      s"Update the cores constant to ${Runtime.getRuntime.availableProcessors}"
+    )
+
+  def benchmarkPingPongActors(numMessagesPerActorPair: Int, numActors: Int, dispatcher: String, throughPut: Int, shutdownTimeout: Duration)(implicit system: ActorSystem): Unit = {
+    val numPairs = numActors / 2
+    val totalNumMessages = numPairs * numMessagesPerActorPair
+    val (actors, latch) = startPingPongActorPairs(numMessagesPerActorPair, numPairs, dispatcher)
+    val startNanoTime = System.nanoTime()
+    initiatePingPongForPairs(actors, inFlight = throughPut * 2)
+    latch.await(shutdownTimeout.toSeconds, TimeUnit.SECONDS)
+    printProgress(totalNumMessages, numActors, startNanoTime)
+  }
+
+  def benchmarkUserServiceActor(numQueriesPerActor: Int, numUserServiceActors: Int, numUsersInDB: Int, randomSeed: Int, dispatcher: String, shutdownTimeout: Duration)(implicit system: ActorSystem): Unit = {
+    val randomGen = new Random(randomSeed)
+    val totalNumberOfQueries = numQueriesPerActor * numUserServiceActors
+    val (actors, latch) = startUserServiceActors(numUserServiceActors, numUsersInDB, numQueriesPerActor, dispatcher, randomGen)
+    val startNanoTime = System.nanoTime()
+    for {
+      serviceActor <- actors
+      _ <- 0 until numQueriesPerActor
+    } {
+      serviceActor ! Request(randomGen.nextInt(numUsersInDB))
+    }
```

Comment:
yeah, it's difficult. One could create threads in the benchmark to have more sender threads, but it's also about flow control. We don't wan't to fill mailboxes with thousands of messages. That's why initially injecting a limited number of tokens (messages) and then pass those around is a simple approach. Perhaps the cross-sending is more realistic than pairs, and you could use that for the stateful queries also? https://github.com/akka/akka/pull/23167/files#diff-abeefc1cdff701f0a633996ccda8dc2d